## Individual Project
Name: Vivian Lim Bi Fang

ID: P101458


# Machine Learning

In [1]:
# import data 
import pandas as pd

# hotel_reviews_data is being preprocess by removing stopwords, punctuation and numbers in assignment 2.
hotel_reviews_data = pd.read_csv("assignment4.csv")
hotel_reviews_data.head()

Lemmatized Sentiment
0  Service friendly staff make us comfortable fam...  Positive
1  family stay one night find hotel much satisfac...  Positive
2  Midsized room clean Good breakfast food option...  Positive
3  Hotel top shop Pavillion Lot suitable couple w...  Positive
4  comfortable accommodation penang hard rock com...  Positive

In [8]:
# split the data into training and test set 

from sklearn.model_selection import train_test_split

input_data = hotel_reviews_data["Lemmatized"]
label_data = hotel_reviews_data["Sentiment"]

x_train, x_test, y_train, y_test = train_test_split(input_data, label_data, test_size = 0.2, random_state=42)
print (len(x_train))
print (len(x_test))
print (len(y_train))
print (len(y_test))

78
20
78
20


## Bag of words (BOW)

In [9]:
from sklearn.feature_extraction.text import CountVectorizer

BOW = CountVectorizer()
BOW.fit(x_train)
BOW_train = BOW.transform(x_train)
# print (BOW_train)
BOW_test = BOW.transform(x_test)

print ((BOW_train.shape[1]))
print ((BOW_test.shape[1]))


735
735


### Classification_BOW

In [10]:
from sklearn.naive_bayes import MultinomialNB
from sklearn.linear_model import LogisticRegression
from sklearn import svm
from sklearn.metrics import classification_report

# naive bayes
NB_BOW = MultinomialNB()
NB_BOW.fit(BOW_train, y_train)
predict_NB_BOW = NB_BOW.predict(BOW_test)
result_NB_BOW = classification_report(predict_NB_BOW, y_test)
print("NB Classifier result:")
print (result_NB_BOW)

# SVM
SVM_BOW = svm.SVC(gamma= "scale")
SVM_BOW.fit(BOW_train, y_train)
predict_SVM_BOW = SVM_BOW.predict(BOW_test)
result_SVM_BOW = classification_report(predict_SVM_BOW, y_test)
print("SVM Classifier result:")
print (result_SVM_BOW)

# Logistic Regression
LR_BOW = LogisticRegression(n_jobs=-1, random_state=42)
LR_BOW.fit(BOW_train, y_train)
predict_LR_BOW = LR_BOW.predict(BOW_test)
result_LR_BOW = classification_report(predict_LR_BOW, y_test)
print("LR Classifier result:")
print (result_LR_BOW)


NB Classifier result:
              precision    recall  f1-score   support

    Negative       0.88      1.00      0.93         7
    Positive       1.00      0.92      0.96        13

    accuracy                           0.95        20
   macro avg       0.94      0.96      0.95        20
weighted avg       0.96      0.95      0.95        20

SVM Classifier result:
              precision    recall  f1-score   support

    Negative       0.75      0.86      0.80         7
    Positive       0.92      0.85      0.88        13

    accuracy                           0.85        20
   macro avg       0.83      0.85      0.84        20
weighted avg       0.86      0.85      0.85        20

LR Classifier result:
              precision    recall  f1-score   support

    Negative       0.75      1.00      0.86         6
    Positive       1.00      0.86      0.92        14

    accuracy                           0.90        20
   macro avg       0.88      0.93      0.89        20
weighte

C:\Users\vivia\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\vivia\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:1544: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 8.
  " = {}.".format(effective_n_jobs(self.n_jobs)))


In [11]:
# from sklearn.model_selection import GridSearchCV

# solver_list = ['liblinear', 'newton-cg', 'lbfgs', 'sag', 'saga']
# params = dict(solver=solver_list)
# log_reg = LogisticRegression(C=1, n_jobs=-1, random_state=34)
# clf = GridSearchCV(log_reg, params, cv=5)
# clf.fit(BOW_train, y_train)
# scores = clf.cv_results_['mean_test_score']

# for score, solver in zip(scores, solver_list):
#     print(f"  {solver} {score:.3f}" )

## TF-IDF

In [12]:
from sklearn.feature_extraction.text import TfidfVectorizer

TFIDF = TfidfVectorizer(analyzer = "word")
TFIDF.fit (x_train)
TFIDF_train = TFIDF.transform (x_train)
TFIDF_test = TFIDF.transform (x_test)

# word_TFIDF = pd.DataFrame(TFIDF_train.toarray(), columns = TFIDF.get_feature_names())
# word_TFIDF

### classification_TF-IDF

In [13]:
# naive bayes
NB_TFIDF = MultinomialNB()
NB_TFIDF.fit(TFIDF_train, y_train)
predict_NB_TFIDF = NB_TFIDF.predict(TFIDF_test)
result_NB_TFIDF = classification_report(predict_NB_TFIDF, y_test)
print("NB Classifier result:")
print (result_NB_TFIDF)

# SVM
SVM_TFIDF = svm.SVC(gamma= "scale")
SVM_TFIDF.fit(TFIDF_train, y_train)
predict_SVM_TFIDF = SVM_TFIDF.predict(TFIDF_test)
result_SVM_TFIDF = classification_report(predict_SVM_TFIDF, y_test)
print("SVM Classifier result:")
print (result_SVM_TFIDF)

# Logistic Regression
LR_TFIDF = LogisticRegression(n_jobs=-1, random_state=42)
LR_TFIDF.fit(TFIDF_train, y_train)
predict_LR_TFIDF = LR_BOW.predict(TFIDF_test)
result_LR_TFIDF = classification_report(predict_LR_TFIDF, y_test)
print("LR Classifier result:")
print (result_LR_TFIDF)

NB Classifier result:
              precision    recall  f1-score   support

    Negative       0.88      1.00      0.93         7
    Positive       1.00      0.92      0.96        13

    accuracy                           0.95        20
   macro avg       0.94      0.96      0.95        20
weighted avg       0.96      0.95      0.95        20

SVM Classifier result:
              precision    recall  f1-score   support

    Negative       0.88      1.00      0.93         7
    Positive       1.00      0.92      0.96        13

    accuracy                           0.95        20
   macro avg       0.94      0.96      0.95        20
weighted avg       0.96      0.95      0.95        20

LR Classifier result:
              precision    recall  f1-score   support

    Negative       0.88      0.70      0.78        10
    Positive       0.75      0.90      0.82        10

    accuracy                           0.80        20
   macro avg       0.81      0.80      0.80        20
weighte

C:\Users\vivia\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:432: FutureWarning: Default solver will be changed to 'lbfgs' in 0.22. Specify a solver to silence this warning.
  FutureWarning)
C:\Users\vivia\Anaconda3\lib\site-packages\sklearn\linear_model\logistic.py:1544: UserWarning: 'n_jobs' > 1 does not have any effect when 'solver' is set to 'liblinear'. Got 'n_jobs' = 8.
  " = {}.".format(effective_n_jobs(self.n_jobs)))


## Doc2vec

In [14]:
from gensim.models import Doc2Vec
from gensim.models.doc2vec import TaggedDocument
import nltk
nltk.download('punkt')
from sklearn import utils


[nltk_data] Downloading package punkt to
[nltk_data]     C:\Users\vivia\AppData\Roaming\nltk_data...
[nltk_data]   Package punkt is already up-to-date!


In [15]:
Doc2vec_data = hotel_reviews_data[["Lemmatized", "Sentiment"]]
train, test = train_test_split(Doc2vec_data, test_size = 0.2, random_state=42)

# print (len(train))
# print (len(test))

In [16]:
def tokenize_text(sentence):
    tokens = []
#     print (sentence)
#     for sent in nltk.sent_tokenize(sentence):
#         print (sent)
    for word in nltk.word_tokenize(sentence):
#         print (word)
#         print (len(word))
        if len(word) < 2:
            continue
        tokens.append(word.lower())
#         print (tokens)
    return tokens

# tagged the documents with unique ID
train_tagged = train.apply(lambda x : TaggedDocument(words = tokenize_text(x["Lemmatized"]),tags=[x.Sentiment]), axis=1)
print (train_tagged)
test_tagged = test.apply(lambda x : TaggedDocument(words = tokenize_text(x["Lemmatized"]),tags=[x.Sentiment]), axis=1)

49    ([strategic, point, locations, shop, provide, ...
70    ([accommodation, interest, not, take, care, fa...
68    ([breakfast, food, delicious, not, satisfactor...
15    ([hotel, heart, metropolitan, traffic, flow, a...
39    ([service, quite, good, satisfy, carriage, air...
                            ...                        
60    ([comfortable, hotel, business, daily, work, h...
71    ([theme, park, vicinity, condo, package, not, ...
14    ([room, clean, comfortable, friendly, staff, s...
92    ([airconditioning, not, work, well, disappoint...
51    ([opinion, service, accommodation, facilities,...
Length: 78, dtype: object


In [17]:
# build vocab

model_dbow = Doc2Vec(dm=0, vector_size=300, negative =5, hs=0, min_count =1, sample=0, workers = 8, alpha=0.025, min_alpha=0.001)
model_dbow.build_vocab([x for x in train_tagged])

# train model

for epoch in range (50):
    model_dbow.train(utils.shuffle([x for x in train_tagged]), total_examples = len(train_tagged.values), epochs =1)
    model_dbow.alpha -= 0.002
    model_dbow.min_alpha = model_dbow.alpha
    
# create final vector feature

def vec_for_learning(model, tagged_docs):
    sents = tagged_docs
    targets, regressors = zip(*[(doc.tags[0], model.infer_vector(doc.words, steps=20)) for doc in sents])
#     print (targets, regressors)
    return targets, regressors

Doc2vec_y_train, Doc2vec_x_train = vec_for_learning(model_dbow, train_tagged)
# print (Doc2Vec_x_train)
Doc2vec_y_test, Doc2vec_x_test = vec_for_learning(model_dbow, test_tagged)

### Classification_Doc2vec

In [18]:
# naive bayes
from sklearn.naive_bayes import GaussianNB

NB_Doc2vec = GaussianNB()
NB_Doc2vec.fit(Doc2vec_x_train, Doc2vec_y_train)
predict_NB_Doc2vec = NB_Doc2vec.predict(Doc2vec_x_test)
result_NB_Doc2vec = classification_report(predict_NB_Doc2vec, Doc2vec_y_test)
print("NB Classifier result:")
print (result_NB_Doc2vec)

# SVM
SVM_Doc2vec = svm.SVC(gamma= "scale")
SVM_Doc2vec.fit(Doc2vec_x_train, Doc2vec_y_train)
predict_SVM_Doc2vec = SVM_Doc2vec.predict(Doc2vec_x_test)
result_SVM_Doc2vec = classification_report(predict_SVM_Doc2vec, Doc2vec_y_test)
print("SVM Classifier result:")
print (result_SVM_Doc2vec)

# Logistic Regression
LR_Doc2vec = svm.SVC(gamma= "scale")
LR_Doc2vec.fit(Doc2vec_x_train, Doc2vec_y_train)
predict_LR_Doc2vec = LR_Doc2vec.predict(Doc2vec_x_test)
result_LR_Doc2vec = classification_report(predict_LR_Doc2vec,  Doc2vec_y_test)
print("LR Classifier result:")
print (result_LR_Doc2vec)

NB Classifier result:
              precision    recall  f1-score   support

    Negative       0.50      0.50      0.50         8
    Positive       0.67      0.67      0.67        12

    accuracy                           0.60        20
   macro avg       0.58      0.58      0.58        20
weighted avg       0.60      0.60      0.60        20

SVM Classifier result:
              precision    recall  f1-score   support

    Negative       0.38      0.60      0.46         5
    Positive       0.83      0.67      0.74        15

    accuracy                           0.65        20
   macro avg       0.60      0.63      0.60        20
weighted avg       0.72      0.65      0.67        20

LR Classifier result:
              precision    recall  f1-score   support

    Negative       0.38      0.60      0.46         5
    Positive       0.83      0.67      0.74        15

    accuracy                           0.65        20
   macro avg       0.60      0.63      0.60        20
weighte

### Conclusion

In Word2vec,

1) NaiveBayes shows the highest precision, recall and f1-score according to the classification report. However the f1-score of the other classifiers (SVM % Logistic Regression) is considered high which is 0.85 and 0.90 respectively despite the value is lower than NaiveBayes.

In TF-IDF,

1) Due to the function of neglecting low weight (high frequency words), NaiveBayes and SVM Classifier give the same and highest acccurcy. However, the f1-score in Logistic Regression if compare with word2vec method.

In Doc2Vec,

1) Doc2vec has the lowest accuracy in terms of  precision, recall and f1-score by comparing with Word2Vec and TF-IDF methods. 




**In short, hotel reviews dataset is recommended to use TF-IDF (either NaiveBayes or LogisticRegression) method for sentiment study.**

## Lexicon Sentiment Analysis

In [2]:
import pandas as pd

In [3]:
# import +ve dataset

mydata1 = pd.read_csv("Hotel +ve reviews.csv")
mydata1.head()

Malay Comments  \
0  Layanan staf amat mesra menjadikan kami sekelu...   
1  kami sekeluarga menginap untuk semalam dan men...   
2  Bilik sederhana besar tetapi bersih. Sarapan p...   
3   Hotel berada di lokasi membeli belah terkemuk...   
4  saya selesa dengan penginapan di hard rock pen...   

                                    Google Translate  \
0  Service is very friendly staff made ??us very ...   
1  Our family stayed one night and found this hot...   
2  Mid-sized room but clean. Good breakfast food ...   
3  Hotel is on the top of a shopping Pavillion, L...   
4  I am comfortable with accommodation in penang ...   

                                      Bing Translate  
0  "Staff very friendly Service made us family ve...  
1  would our family for today and found this hote...  
2  Simple rooms but clean. Breakfast a nice large...  
3  Hotel located in leading shopping, Pavillion, ...  
4  I'm comfortable with a stay at the hard rock p...

In [4]:
# import -ve dataset

mydata2 = pd.read_csv("Hotel -ve reviews.csv")
mydata2.head()

Malay Comments  \
0  Pada pendapat saya, segala perkhidmatan dan ke...   
1  Saya di lantai 3 bersama ibu saya. Selain kele...   
2   kami diberi bilik hotel di tower 2, level 1 (...   
3  Harga mahal, bilik tak selesa, kebersihan patu...   
4  Kehilangan 2 bentuk cincin perkahwinan yang di...   

                                    Google Translate  \
0  In my opinion, all the services and accommodat...   
1  I was on the 3rd floor with my mother. In addi...   
2  we were given a room in tower 2, level 1 (at t...   
3  The price is expensive, uncomfortable rooms, c...   
4  Losing two wedding rings and stored in bags an...   

                                      Bing Translate  
0  In my opinion, services and accommodation is a...  
1  I am on the 3rd floor with my mother. In addit...  
2  We were given hotel rooms in tower 2, level 1 ...  
3  "The price of expensive, rooms are comfortable...  
4  Loss of 2 form of wedding rings that are store...

In [5]:
# merge +ve&-ve files 

mydata = mydata1.append(mydata2).reset_index(drop=True)
pd.set_option("display.max_rows", None)
mydata

Malay Comments  \
0   Layanan staf amat mesra menjadikan kami sekelu...   
1   kami sekeluarga menginap untuk semalam dan men...   
2   Bilik sederhana besar tetapi bersih. Sarapan p...   
3    Hotel berada di lokasi membeli belah terkemuk...   
4   saya selesa dengan penginapan di hard rock pen...   
5   Percutian yang menarik di hotel yang bersih, b...   
6   Penginapan yang bertaraf 5 bintang. Layanan ka...   
7   Sangat selesa..berdekatan dengan pantai..sesua...   
8   Sebuah hotel mewah yang terletak di hujung pul...   
9   Sepanjang penginapan kami, kami sangat berpuas...   
10  Hotel yang begitu istimewa dan begitu mengagum...   
11      Saya sangat berpuas hati dengan layanan ya...   
12      saya tidak pernah pergi ke Pulau Pinang se...   
13   bilik bersih dan rekaan dalam menarik, kakita...   
14  bilik bersih, selesa dan pekerja mesra, memuas...   
15  Hotel di tengah bandar metropolitan yang agak ...   
16  Hotel ini berdekatan dengan penang times squar...   
17  hotel yang bagus untuk menginap sekiranya tida...   
18  Menyeronokan dan berpuashati dengan persekitar...   
19  Menyeronokkan dan berpuashati dengan persekita...   
20  cuaca yang sejuk, bilik yang selesa, tandas ya...   
21  Penginapan yang sangat seronok, bilik yang lua...   
22  TEMPAT NI AMAT BESAR BILIKNYA ..UDARA NYAMAN S...   
23  sangat selesa...sesuai untuk mereka yang sukak...   
24  sangat berhampiran dengan taman tema luar dan ...   
25  sederhana sahaja,namun sangat selesa mungkin k...   
26  Theme Park Hotel menyediakan penginapan yang s...   
27  servis di kaunter yang baik dan pantas, bilik ...   
28  Satu pengalaman yang baik tinggal di sini, han...   
29  melalui agoda saya membuat tempahan dan bila b...   
30  KAMI MENYUKAI PENGINAPAN KAMI KERANA ANDA SANG...   
31  Bilik yang sangat luas dengan katil yang besar...   
32  Sungguh selesa. Ruang bilik yang cukup luas da...   
33  hotel ini merupakan hotel yang terbaik bagi sa...   
34  Saya tidak tahan sejuk terutama waktu malam da...   
35  Sangat memuaskan, walaupun bangunan hotel lama...   
36  perfect..ruang bilik dan cafe yang cukup menar...   
37  bilik yang baru dan bersih, lokasi di pusat ba...   
38  Kedudukan strategik untuk kemana2. Untuk menja...   
39  perkhidmatan agak baik dan saya berpuas hati d...   
40  penginapan yang sangat memuaskan. Apa yang ter...   
41  memang cantik permandangan beach dekat depan b...   
42  Tempat strategik dan sangat memudahkan kami be...   
43  Kedudukan hotel yang strategik untuk mereka ya...   
44  Amat selesa, nyaman, pemandangan yang menarik,...   
45  Hotel ini hanya bertaraf 2 bintang. Sesuai utk...   
46  penginapan saya dan suami bersama 3 anak kecil...   
47   Mudah untuk mendapatkan kemudahan kedai runci...   
48  Bilik yang disediakan selesa dan luas. Keadaan...   
49   Tempat yang strategik dengan lokasi-lokasi un...   
50  Staff yang ramah , hotel yang bersih , dan jug...   
51  Pada pendapat saya, segala perkhidmatan dan ke...   
52  Saya di lantai 3 bersama ibu saya. Selain kele...   
53   kami diberi bilik hotel di tower 2, level 1 (...   
54  Harga mahal, bilik tak selesa, kebersihan patu...   
55  Kehilangan 2 bentuk cincin perkahwinan yang di...   
56  penginapan bagus tapi bilik kali ini sangat ti...   
57  Saya amat berpuas hati dengan persekitaran hot...   
58  Breakfast harga mahal perkerja sombong,parking...   
59  kolam renang tidak boleh digunakan menyebabkan...   
60  Hotel yang selesa untuk urusan perniagaan dan ...   
61  Mula-mula sampai nak check in dah mendapat lay...   
62  penginapan di First World hotel tidak mempunya...   
63   Hotel di puncak Genting yang mempunyai suasan...   
64  bilik yang ditawarkan sangat cantik, sesuai un...   
65  sepanjang penginapan,kami hanya selesa berada ...   
66  Check in pantas dan senang tapi tiada free wif...   
67  Bilik perlu ada penyelenggaraan yang berterusa...   
68  makanan sarapan tidak sedap, tidak memuaskan. ...   
69  Lokasi yang sesuai untuk pelancong mendapatka

In [6]:
newdata = mydata.drop(["Malay Comments", "Bing Translate"], axis=1)
newdata.head()


Google Translate
0  Service is very friendly staff made ??us very ...
1  Our family stayed one night and found this hot...
2  Mid-sized room but clean. Good breakfast food ...
3  Hotel is on the top of a shopping Pavillion, L...
4  I am comfortable with accommodation in penang ...

In [7]:
#change column name
Column_name = ['Google_Translate']
newdata.columns = Column_name
newdata.head()

Google_Translate
0  Service is very friendly staff made ??us very ...
1  Our family stayed one night and found this hot...
2  Mid-sized room but clean. Good breakfast food ...
3  Hotel is on the top of a shopping Pavillion, L...
4  I am comfortable with accommodation in penang ...

In [8]:
# create checkpoint before preprocess 

newdata2 = newdata

In [9]:
# preprocess functions to remove stopwords & punctuation & numbers

import string 
from nltk.corpus import stopwords
import re

new_stopwords = set(stopwords.words("english")) - set(["not"])
# print(new_stopwords)

def text_process(x):
    
    #no_punc = []    
    nopunc = [char for char in x if char not in string.punctuation]
    nopunc = "".join(nopunc)
    
    no_punc = [char for char in nopunc.split() if char.lower() not in new_stopwords]
    return no_punc
    
    


def remove_numbers(x):
    
    no_punc = text_process(x)
    reviews = reviews = re.sub("[^a-zA-Z]", " ", str(no_punc))
    return reviews    
   

In [10]:
# preprocess 

clean_text = newdata2["Google_Translate"].apply(remove_numbers)
newdata2["Clean_Text"] =pd.DataFrame(clean_text)
newdata2.head()

Google_Translate  \
0  Service is very friendly staff made ??us very ...   
1  Our family stayed one night and found this hot...   
2  Mid-sized room but clean. Good breakfast food ...   
3  Hotel is on the top of a shopping Pavillion, L...   
4  I am comfortable with accommodation in penang ...   

                                          Clean_Text  
0    Service    friendly    staff    made    us  ...  
1    family    stayed    one    night    found   ...  
2    Midsized    room    clean    Good    breakfa...  
3    Hotel    top    shopping    Pavillion    Lot...  
4    comfortable    accommodation    penang    ha...

In [11]:
#create another checkpoint 

newdata3 = newdata2
newdata3.head()

Google_Translate  \
0  Service is very friendly staff made ??us very ...   
1  Our family stayed one night and found this hot...   
2  Mid-sized room but clean. Good breakfast food ...   
3  Hotel is on the top of a shopping Pavillion, L...   
4  I am comfortable with accommodation in penang ...   

                                          Clean_Text  
0    Service    friendly    staff    made    us  ...  
1    family    stayed    one    night    found   ...  
2    Midsized    room    clean    Good    breakfa...  
3    Hotel    top    shopping    Pavillion    Lot...  
4    comfortable    accommodation    penang    ha...

In [12]:
# insert sentiment labels for accuracy test later on [label comparison] 

newdata3.loc[newdata.index[0:51], "Sentiment"] = ["Positive"]
newdata3.loc[newdata.index[51:], "Sentiment"] = ["Negative"]
newdata3

Google_Translate  \
0   Service is very friendly staff made ??us very ...   
1   Our family stayed one night and found this hot...   
2   Mid-sized room but clean. Good breakfast food ...   
3   Hotel is on the top of a shopping Pavillion, L...   
4   I am comfortable with accommodation in penang ...   
5   Fantastic holiday in this clean, large room, s...   
6   5-star accommodation. Staff are great, have al...   
7   Very comfortable .. close to the beach .. perf...   
8   A luxury hotel located at the end pulau.makana...   
9   During our stay, we were very pleased with the...   
10  Hotels are so special and so amazing decor and...   
11  I am very satisfied with the service provided ...   
12  I never went to Penang before. This is my firs...   
13  the rooms are clean and attractive designs in,...   
14  the rooms are clean, comfortable and friendly ...   
15  Hotel in the heart of metropolitan traffic flo...   
16  The hotel is close to Penang Times Square. At ...   
17  a nice hotel to stay if they do not have a veh...   
18  Exciting and are satisfied with the environmen...   
19  Relaxed and satisfied with the hotel environme...   
20  cold weather, comfortable, clean restrooms, ca...   
21  Accommodation is very fun, spacious than expec...   
22  VERY BIG PLACE NI balmy room .. DELICIOUS DELI...   
23  very comfortable ... perfect for those who lov...   
24  very close to the outdoor theme park and the r...   
25  mild, but very uncomfortable probably because ...   
26  Theme Park Hotel provides comfortable accommod...   
27  Good counter service and fast, spacious and co...   
28  A good experience to stay here, just some chin...   
29  I made a booking through Agoda, and when there...   
30  ACCOMMODATION WE LIKE OUR WORTH SEEING YOU AT ...   
31  A very spacious room with a bed in the garden ...   
32  It's comfortable. Living rooms are quite spaci...   
33  This hotel is the best hotel for me so far and...   
34  I can not stand the cold, especially at night ...   
35  Very satisfactory, although the old hotel buil...   
36  perfect .. living room and cafe enough interes...   
37  rooms are new and clean, the location in the c...   
38  Strategic position to kemana2. For dining, wal...   
39  service is quite good and I am satisfied with ...   
40  very satisfying stay. What's best is its locat...   
41  really beautiful scenery near the beach front ...   
42  Strategic and very easy for us to walk, enjoy ...   
43  The hotel's strategic position for those who p...   
44  Extremely comfortable, cozy, attractive views,...   
45  It is a 2-star. Suitable for medium budget tra...   
46  stay with my husband and three small children ...   
47  Easy to get a grocery facility. Easy for break...   
48  The rooms are comfortable and spacious availab...   
49  Strategic point of locations for shopping and ...   
50  Staff are friendly, clean hotel, and also a gr...   
51  In my opinion, all the services and accommodat...   
52  I was on the 3rd floor with my mother. In addi...   
53  we were given a room in tower 2, level 1 (at t...   
54  The price is expensive, uncomfortable rooms, c...   
55  Losing two wedding rings and stored in bags an...   
56  stay was nice but the room is very uncomfortab...   
57  I am very satisfied with the hotel, but poor s...   
58  Breakfast expensive price arrogant workers, pa...   
59  pool can not be used cause my son complained a...   
60  This comfortable hotel for both business and d...   
61  First want to check in've got less friendly se...   
62  accommodation at First World Hotel does not ha...   
63  At the peak of Genting Hotel has a peaceful at...   
64  all are very beautiful, perfect for the night....   
65  relaxation, we just stay in the room. lobby sp...   
66  Check in is quick and easy, but there is no fr...   
67  Rooms have no ongoing maintenance. Bath tubs s...   
68  breakfast food is delicious, is not satisfacto...   
69  The location is ideal for tourists to get a

In [17]:
#import lexicon file 

lexicon_sentiment = pd.read_csv("lexicon_sentiment_file_edited.csv")
lexicon_sentiment.head()

ADJ_Text  Score
0  accessible      1
1      active      1
2        aged      1
3     amazing      1
4    arrogant     -1

In [18]:
#Calculate sum score & sentiment 

def sentiment(sentence, score):
    print ("Sentence" + sentence)
    print ("Score:" + str(score))
    if (score > 0):
        print('Review sentiment: POSITIVE')
        senti = 'Positive'
    elif (score < 0):
        print('Review sentiment: NEGATIVE')
        senti = 'Negative'
    else:
        print('Review sentiment: NEUTRAL')
        senti = 'Neutral'
    print ("")
    return senti



def whole_reviews(review):
    sentence_array = []
    score_array = []
    sentiment_array = []
    
    for sentences in newdata3.Clean_Text:
        score = 0

        for word in sentences.split():

            if word in lexicon_sentiment["ADJ_Text"].values:
                score += lexicon_sentiment.query("ADJ_Text == @word")["Score"].values[0]


        sentence_array.append(sentences)
        score_array.append(score)
        senti = sentiment(sentences, score)
        sentiment_array.append(senti)
    return sentence_array, score_array, sentiment_array
    
    

sentence, score, senti = whole_reviews(newdata3["Clean_Text"])

Sentence  Service    friendly    staff    made    us    comfortable    family    stay    Decoration    hotel    interesting    quite    different    hotel    occupancy    hotel    ever    Services    relaxing    spa    experience    Hotel    located    close    klcc    walk    rent    buggy    also    interesting    experience    Tip    Puller    Attention    Sights    select    room    facing    klcc    interesting    views    seen    night    see    light    spill    klcc  
Score:7
Review sentiment: POSITIVE

Sentence  family    stayed    one    night    found    hotel    much    satisfaction    near    klcc    allow    family    go    enjoy    lunch    dinner    ease    booked    rooms    facing    klcc    give    beautiful    views    especially    sunset  
Score:1
Review sentiment: POSITIVE

Sentence  Midsized    room    clean    Good    breakfast    food    options    Places    entertainment    ground    floor    hotel    Hard    Rock    Cafe    Kuala    Lumpur    Menara    Kuala

Sentence  not    stand    cold    especially    night    early    pagiTapi    loves    Location    located    building    Gentingkaunter    Theme    Park    tickets    theme    park    located    entrance    hotel    lobby    iniRuang    selesaMudah    eat    buy    general    requirements    Handphone    Topup    Card  
Score:0
Review sentiment: NEUTRAL

Sentence  satisfactory    although    old    hotel    building    modern    interior    design    makes    comfortable    staff    friendly    quite    helpful    location    close    business    office    business    near    center    local    food  
Score:6
Review sentiment: POSITIVE

Sentence  perfect    living    room    cafe    enough    interest    nearby    shopping    complex    restaurant    public    transport    also    readily    available  
Score:4
Review sentiment: POSITIVE

Sentence  rooms    new    clean    location    city    center    near    shopping    store    ipoh    parade    also    close    dining    places   

Sentence  Employees    friendly    quiet    surroundings    lobby    condo    jugak    unattractive    gloomy    nothing    exciting  
Score:-2
Review sentiment: NEGATIVE

Sentence  parking    problems    information    available    hotel    not    clearly    known    relatively    secluded    location  
Score:0
Review sentiment: NEUTRAL

Sentence  quite    disappointing    door    damaged    maintenance    not    tell    door    not    opened    inside    imagine    fire    get    locked    room    changed    rooms    dangerous  
Score:-4
Review sentiment: NEGATIVE

Sentence  expect    hotel    apart    pretty    comfortable    however    bit    disappointed    hotel    room    small    equivalent    budget    hotels    like    tune    inadequate    money    spend    certainly    would    not    stay  
Score:-3
Review sentiment: NEGATIVE

Sentence  normal    accommodation    room    small    comfortable    hotel    provides    breakfast    bit    disappointing    plain    bread    noo

In [19]:
# create a data frame for accuracy test later on.

lexicon_result = pd.DataFrame()
lexicon_result['Sentence'] = pd.Series(sentence).values
lexicon_result['Score'] = pd.Series(score).values
lexicon_result['Sentiment'] = pd.Series(senti).values

lexicon_result


Sentence  Score Sentiment
0     Service    friendly    staff    made    us  ...      7  Positive
1     family    stayed    one    night    found   ...      1  Positive
2     Midsized    room    clean    Good    breakfa...      4  Positive
3     Hotel    top    shopping    Pavillion    Lot...      3  Positive
4     comfortable    accommodation    penang    ha...      5  Positive
5     Fantastic    holiday    clean    large    ro...      2  Positive
6      star    accommodation    Staff    great    ...      6  Positive
7     comfortable    close    beach    perfect    ...      3  Positive
8     luxury    hotel    located    end    pulauma...     -1  Negative
9     stay    pleased    hotel    atmosphere    ex...      4  Positive
10    Hotels    special    amazing    decor    lay...      6  Positive
11    satisfied    service    provided    staff   ...      7  Positive
12    never    went    Penang    first    trip    ...      7  Positive
13    rooms    clean    attractive    designs    s...     11  Positive
14    rooms    clean    comfortable    friendly   ...      5  Positive
15    Hotel    heart    metropolitan    traffic   ...      3  Positive
16    hotel    close    Penang    Times    Square ...      0   Neutral
17    nice    hotel    stay    not    vehicle    h...      2  Positive
18    Exciting    satisfied    environment    hote...      4  Positive
19    Relaxed    satisfied    hotel    environment...      1  Positive
20    cold    weather    comfortable    clean    r...      2  Positive
21    Accommodation    fun    spacious    expected...      1  Positive
22    BIG    PLACE    NI    balmy    room    DELIC...      1  Positive
23    comfortable    perfect    love    excitement...      2  Positive
24    close    outdoor    theme    park    rooms  ...      2  Positive
25    mild    uncomfortable    probably    hotel  ...     -1  Negative
26    Theme    Park    Hotel    provides    comfor...      7  Positive
27    Good    counter    service    fast    spacio...      5  Positive
28    good    experience    stay    chinese    res...      4  Positive
29    made    booking    Agoda    rooms    availab...      6  Positive
30    ACCOMMODATION    LIKE    WORTH    SEEING    ...      0   Neutral
31    spacious    room    bed    garden    besarbe...      5  Positive
32    comfortable    Living    rooms    quite    s...      4  Positive
33    hotel    best    hotel    far    soul    ini...      1  Positive
34    not    stand    cold    especially    night ...      0   Neutral
35    satisfactory    although    old    hotel    ...      6  Positive
36    perfect    living    room    cafe    enough ...      4  Positive
37    rooms    new    clean    location    city   ...      2  Positive
38    Strategic    position    kemana     dining  ...      1  Positive
39    service    quite    good    satisfied    car...      2  Positive
40    satisfying    stay    Whats    best    locat...      3  Positive
41    really    beautiful    scenery    near    be...      2  Positive
42    Strategic    easy    us    walk    enjoy    ...      4  Positive
43    hotels    strategic    position    prefer   ...      2  Positive
44    Extremely    comfortable    cozy    attracti...      2  Positive
45     star    Suitable    medium    budget    tra...      2  Positive
46    stay    husband    three    small    childre...      3  Positive
47    Easy    get    grocery    facility    Easy  ...      3  Positive
48    rooms    comfortable    spacious    availabl...      6  Positive
49    Strategic    point    locations    shopping ...      2  Positive
50    Staff    friendly    clean    hotel    also ...      3  Positive
51    opinion    services    accommodation    faci...      0   Neutral
52     rd    floor    mother    addition    advant...     -4  Negative
53    given    room    tower         level        ...     -4  Negative
54    price    expensive    uncomfortable    rooms...     -2  Negative
55    Losing    two    wedding    rings    stored ...      0   Neutra

In [20]:
# accuracy test 
import numpy as np

#add the sentiment column from newdata3 dataframe into lexicon_result dataframe
lexicon_result['Original_Sentiment'] = newdata3['Sentiment'] 
# lexicon_result.head()

lexicon_result["Sentiment_Match"] = np.where(lexicon_result["Sentiment"] == newdata3["Sentiment"], "True", "False")
lexicon_result


Sentence  Score Sentiment  \
0     Service    friendly    staff    made    us  ...      7  Positive   
1     family    stayed    one    night    found   ...      1  Positive   
2     Midsized    room    clean    Good    breakfa...      4  Positive   
3     Hotel    top    shopping    Pavillion    Lot...      3  Positive   
4     comfortable    accommodation    penang    ha...      5  Positive   
5     Fantastic    holiday    clean    large    ro...      2  Positive   
6      star    accommodation    Staff    great    ...      6  Positive   
7     comfortable    close    beach    perfect    ...      3  Positive   
8     luxury    hotel    located    end    pulauma...     -1  Negative   
9     stay    pleased    hotel    atmosphere    ex...      4  Positive   
10    Hotels    special    amazing    decor    lay...      6  Positive   
11    satisfied    service    provided    staff   ...      7  Positive   
12    never    went    Penang    first    trip    ...      7  Positive   
13    rooms    clean    attractive    designs    s...     11  Positive   
14    rooms    clean    comfortable    friendly   ...      5  Positive   
15    Hotel    heart    metropolitan    traffic   ...      3  Positive   
16    hotel    close    Penang    Times    Square ...      0   Neutral   
17    nice    hotel    stay    not    vehicle    h...      2  Positive   
18    Exciting    satisfied    environment    hote...      4  Positive   
19    Relaxed    satisfied    hotel    environment...      1  Positive   
20    cold    weather    comfortable    clean    r...      2  Positive   
21    Accommodation    fun    spacious    expected...      1  Positive   
22    BIG    PLACE    NI    balmy    room    DELIC...      1  Positive   
23    comfortable    perfect    love    excitement...      2  Positive   
24    close    outdoor    theme    park    rooms  ...      2  Positive   
25    mild    uncomfortable    probably    hotel  ...     -1  Negative   
26    Theme    Park    Hotel    provides    comfor...      7  Positive   
27    Good    counter    service    fast    spacio...      5  Positive   
28    good    experience    stay    chinese    res...      4  Positive   
29    made    booking    Agoda    rooms    availab...      6  Positive   
30    ACCOMMODATION    LIKE    WORTH    SEEING    ...      0   Neutral   
31    spacious    room    bed    garden    besarbe...      5  Positive   
32    comfortable    Living    rooms    quite    s...      4  Positive   
33    hotel    best    hotel    far    soul    ini...      1  Positive   
34    not    stand    cold    especially    night ...      0   Neutral   
35    satisfactory    although    old    hotel    ...      6  Positive   
36    perfect    living    room    cafe    enough ...      4  Positive   
37    rooms    new    clean    location    city   ...      2  Positive   
38    Strategic    position    kemana     dining  ...      1  Positive   
39    service    quite    good    satisfied    car...      2  Positive   
40    satisfying    stay    Whats    best    locat...      3  Positive   
41    really    beautiful    scenery    near    be...      2  Positive   
42    Strategic    easy    us    walk    enjoy    ...      4  Positive   
43    hotels    strategic    position    prefer   ...      2  Positive   
44    Extremely    comfortable    cozy    attracti...      2  Positive   
45     star    Suitable    medium    budget    tra...      2  Positive   
46    stay    husband    three    small    childre...      3  Positive   
47    Easy    get    grocery    facility    Easy  ...      3  Positive   
48    rooms    comfortable    spacious    availabl...      6  Positive   
49    Strategic    point    locations    shopping ...      2  Positive   
50    Staff    friendly    clean    hotel    also ...      3  Positive   
51    opinion    services    accommodation    faci...      0   Neutral   
52     rd    floor    mother    addition    advant...     -4  Negative   
53    given    room    tower         level       

In [21]:
# get the freq of ["True/False"] values based on original sentiment

result = lexicon_result.groupby('Original_Sentiment').Sentiment_Match.value_counts()
result

Original_Sentiment  Sentiment_Match
Negative            True               36
                    False              11
Positive            True               46
                    False               5
Name: Sentiment_Match, dtype: int64

In [23]:
# calculate percentage of accuracy 

negative_accuracy = '{:.2f}'.format((36/(36+11))*100)
print ("The accuracy of sentiment analysis for negative hotel review is :" + str(negative_accuracy))

positive_accuracy = '{:.2f}'.format((46/(46+5))*100)
print ("The accuracy of sentiment analysis for positive hotel review is :" + str(positive_accuracy))

aver_accuracy = '{:.2f}'.format((((36/(36+11))*100) + ((46/(46+5))*100))/2)
print ("Average accuracy= : " + str(aver_accuracy))


The accuracy of sentiment analysis for negative hotel review is :76.60
The accuracy of sentiment analysis for positive hotel review is :90.20
Average accuracy= : 83.40


# Overall Conclusion


By comparing the effectiveness between both **Machine Learning Classfier** and **Lexicon Analysis**, 
- Machine Learning (Naive Bayes and Logistic) has higher accuracy compare to Lexicon Analysis
- Naive Bayes & Logistic Regression can acheive up to 96% of accuracy whist Lexicon analysis gives an average of 83.40% as the highest accuracy. 
- Hence to summarize, Machine Learning works better for hotel reviews sentiment analysis. 
- Machine Learning is better for this case because good model is created by testing with several parameters. Lexicon analysis method could be improved by adding more rules or parameters to enhance its function. 